cpp/src/arrow/flight/protocol.cc

```diff
@@ -0,0 +1,20 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+
+#include "arrow/flight/protocol.h"
+
+#include "arrow/flight/Flight.grpc.pb.cc"  // NOLINT
```

Comment:
So the trick of `protocol.cc` and `protocol.h` is to make sure you can only get the generated Protobuf code along with our template specialization?

Reply:
Right

Reply:
It appears that the linker (at least with gcc on Linux) was seeing the vtable generated in Flight.grpc.pb.cc and overriding the ones in server.cc/client.cc. This was the only way I could make sure that the same code is generated in all places

Reply:
I think I'd appreciate a comment about that just so that someone else looking at it isn't confused.

Reply:
done